src/Symfony/Bundle/FrameworkBundle/Test/BrowserKitAssertionsTrait.php

```diff
@@ -0,0 +1,159 @@
+    {
+        self::assertThat(self::getResponse(), new ResponseConstraint\ResponseHasCookie($name, $path, $domain), $message);
+    }
+
+    public static function assertResponseNotHasCookie(string $name, string $path = '/', string $domain = null, string $message = ''): void
+    {
+        self::assertThat(self::getResponse(), new LogicalNot(new ResponseConstraint\ResponseHasCookie($name, $path, $domain)), $message);
+    }
+
+    public static function assertResponseCookieValueSame(string $name, string $expectedValue, string $path = '/', string $domain = null, string $message = ''): void
+    {
+        self::assertThat(self::getResponse(), LogicalAnd::fromConstraints(
+            new ResponseConstraint\ResponseHasCookie($name, $path, $domain),
+            new ResponseConstraint\ResponseCookieValueSame($name, $expectedValue, $path, $domain)
+        ), $message);
+    }
+
+    public static function assertBrowserHasCookie(string $name, string $path = '/', string $domain = null, string $message = ''): void
+    {
+        self::assertThat(self::getClient(), new BrowserKitConstraint\BrowserHasCookie($name, $path, $domain), $message);
+    }
+
+    public static function assertBrowserNotHasCookie(string $name, string $path = '/', string $domain = null, string $message = ''): void
+    {
+        self::assertThat(self::getClient(), new LogicalNot(new BrowserKitConstraint\BrowserHasCookie($name, $path, $domain)), $message);
+    }
+
+    public static function assertBrowserCookieValueSame(string $name, string $expectedValue, bool $raw = false, string $path = '/', string $domain = null, string $message = ''): void
+    {
+        self::assertThat(self::getClient(), LogicalAnd::fromConstraints(
+            new BrowserKitConstraint\BrowserHasCookie($name, $path, $domain),
+            new BrowserKitConstraint\BrowserCookieValueSame($name, $expectedValue, $raw, $path, $domain)
+        ), $message);
+    }
+
+    public static function assertRequestAttributeValueSame(string $name, string $expectedValue, string $message = ''): void
+    {
+        self::assertThat(self::getRequest(), new ResponseConstraint\RequestAttributeValueSame($name, $expectedValue), $message);
+    }
+
+    public static function assertRouteSame($expectedRoute, array $parameters = [], string $message = ''): void
+    {
+        $constraint = new ResponseConstraint\RequestAttributeValueSame('_route', $expectedRoute);
+        $constraints = [];
+        foreach ($parameters as $key => $value) {
+            $constraints[] = new ResponseConstraint\RequestAttributeValueSame($key, $value);
+        }
+        if ($constraints) {
+            $constraint = LogicalAnd::fromConstraints($constraint, ...$constraints);
+        }
+
+        self::assertThat(self::getRequest(), $constraint, $message);
+    }
+
+    private static function getClient(AbstractBrowser $newClient = null): ?AbstractBrowser
```

Comment:
Widening the typehint from `KernelBrowser` to `AbstractBrowser` could be considered a tiny BC break in case someone use the trait directly instead of using `WebTestCase`, but it should have done like this from the beginning to allow to use these assertions with Panther, Goutte, and all other BrowserKit based clients.

Reply:
Maybe submit this as bug fix on 4.3, since this is brand new?

Reply:
done